functionary/core/tests/models/test_function.py

```diff
@@ -0,0 +1,77 @@
+import pytest
+from django_celery_beat.models import CrontabSchedule, PeriodicTask
+
+from core.models import Function, Package, ScheduledTask, Task, Team, User
+
+
+@pytest.fixture
+def user():
+    return User.objects.create(username="user")
+
+
+@pytest.fixture
+def team():
+    return Team.objects.create(name="team")
+
+
+@pytest.fixture
+def environment(team):
+    return team.environments.get()
+
+
+@pytest.fixture
+def package(environment):
+    return Package.objects.create(name="testpackage", environment=environment)
+
+
+@pytest.fixture
+def function(package):
+    function_schema = {
+        "title": "test",
+        "type": "object",
+        "properties": {"prop1": {"type": "integer"}},
+    }
+    return Function.objects.create(
+        name="testfunction", package=package, schema=function_schema, active=True
+    )
+
+
+@pytest.fixture
+def task(function, environment, admin_user):
+    return Task.objects.create(
+        function=function,
+        environment=environment,
+        parameters={"prop1": "value1"},
+        creator=admin_user,
+    )
+
+
+@pytest.fixture
+def periodic_task():
+    return PeriodicTask.objects.create(
+        name="periodicTemp",
+        task="task1",
+        crontab=CrontabSchedule.objects.create(hour=7, minute=30, day_of_week=1),
+    )
+
+
+@pytest.fixture
+def scheduled_task(function, environment, user, periodic_task):
+    return ScheduledTask.objects.create(
+        name="testtask",
+        description="description",
+        creator=user,
+        function=function,
+        environment=environment,
+        parameters={"name": "input", "summary": "summary", "type": "text"},
+        periodic_task=periodic_task,
+    )
+
+
+@pytest.mark.django_db
+def test_deactivate(function):
+    """This marks the function to inactive and scheduled tasks to pause"""
+    function.deactivate()
+    assert not function.active
+    for scheduled_t in function.scheduled_tasks.all():
```

Comment:
You've somewhat similarly written this second part of the test in a way that it's easy to trick yourself into thinking it's passing, which you've done.

You're not actually loading the `scheduled_task` fixture.  So when you get here you're iterating over an empty set.  That means you're not actually testing the thing you're intending to test.  Another way you could possibly fool yourself here is if the scheduled tasks were already "PAUSED".  Then you've not verified that the deactivation process is what paused them.

What you can do is an upfront assertion to verify that there are scheduled tasks and they are currently active.  So at the front of your test, do:

```python
    assert function.active is True
    assert function.scheduled_tasks.filter(status=ScheduledTask.ACTIVE).exists()
```

Then you know that the assertions run after you call `function.deactivate()` are meaningful.

Also, note the use of the `ScheduledTask.ACTIVE` constant versus the string "ACTIVE".  Do the same thing for `PAUSED`.

Reply:
Did upfront assertion and used the constant, paused/active